contracts/pol/CarbonPOL.sol

```diff
@@ -0,0 +1,415 @@
+     */
+    function initialize() public initializer {
+        __CarbonPOL_init();
+    }
+
+    // solhint-disable func-name-mixedcase
+
+    /**
+     * @dev initializes the contract and its parents
+     */
+    function __CarbonPOL_init() internal onlyInitializing {
+        __Upgradeable_init();
+        __ReentrancyGuard_init();
+
+        __CarbonPOL_init_unchained();
+    }
+
+    /**
+     * @dev performs contract-specific initialization
+     */
+    function __CarbonPOL_init_unchained() internal onlyInitializing {
+        _setRewardsPPM(2000);
+        // 1% tradeable per day (equal to 1389 ppb per block, given 7200 blocks per day)
+        _setMaxTradeablePPB(1389);
+        // max slippage set to 0.3% initially
+        _setMaxSlippagePPM(3000);
+    }
+
+    /**
+     * @dev authorize the contract to receive the native token
+     */
+    receive() external payable {}
+
+    /**
+     * @dev perform various validations for the token array
+     */
+    modifier validateTokens(Token[] calldata tokens, uint24[] calldata poolFees) {
+        _validateTokens(tokens, poolFees);
+        _;
+    }
+
+    /**
+     * @inheritdoc Upgradeable
+     */
+    function version() public pure override(IVersioned, Upgradeable) returns (uint16) {
+        return 1;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function setRewardsPPM(uint32 newRewardsPPM) external onlyAdmin validFee(newRewardsPPM) {
+        _setRewardsPPM(newRewardsPPM);
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function setMaxSlippagePPM(uint32 newMaxSlippagePPM) external onlyAdmin validFee(newMaxSlippagePPM) {
+        _setMaxSlippagePPM(newMaxSlippagePPM);
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function setMaxTradeablePPB(uint32 newmaxTradeablePPB) external onlyAdmin validValue(newmaxTradeablePPB) {
+        _setMaxTradeablePPB(newmaxTradeablePPB);
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function rewardsPPM() external view returns (uint32) {
+        return _rewardsPPM;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function maxSlippagePPM() external view returns (uint32) {
+        return _maxSlippagePPM;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function maxTradeablePPB() external view returns (uint32) {
+        return _maxTradeablePPBPerBlock;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function lastTradedBlock(Token token) external view returns (uint256) {
+        return _lastTradedBlock[token];
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function maxTradeableAmount(Token token) external view returns (uint256) {
+        uint256 balance = token.balanceOf(address(this));
+        return _maxTradeableAmount(token, balance);
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function tradeTokens(
+        Token[] calldata tokens,
+        uint24[] calldata poolFees
```

Comment:
Consider an array of objects instead, where each object is of type:
```
struct TradeToken {
    Token token,
    uint24 poolFee
}
```

Reply:
Agreed on the approach, seems better, but there's an open discussion about the poolFees currently, so let's decide later.